### DISCUSSION TITLE
customizing where composer autoloader pulls from at runtime?

### DISCUSSION THREAD

**Discussion by dadamssg** [composer/composer#11937] [Q&A / Support]
Let me preface by saying i'm working with a very old php project that does not use composer's autoloader. I'm trying to work out how I can load classes from different directories based on the requested subdomain. The app responds to multiple subdomains.

For example, let's say i have this structure:
```
/src
  /User
    UserRepository.php
    User.php
/subdomain
  /acme
    /User
      User.php
```

If a request comes in from the acme subdomain, `https://acme.domain.com/api/users/1` then the following files should be autoloaded:
```php
// acme does not have its own version of UserRepository, so use core code
src/User/UserRepository.php

// acme *does* have its own User class, so load the custom file
subdomain/acme/User/User.php
```

The namespace will be the same in both `src/User/User.php` and `subdomain/acme/User/User.php`: 
```php
namespace MyCompany\MyProject\User;
```
And this will be in `composer.json`:

```json
"autoload": {
    "psr-4": {
        "MyCompany\\MyProject\\": "src/"
    }
},
```

Do i just need to define my own psr-4 autoloader that checks for custom code and register that _before_ composer's? or is there a way to hook into composer's somehow?

**Comment by stof** [comment]
you could create your own custom instance of the ClassLoader (you can reuse the implementation of the Composer ClassLoader instead of creating your own).
Based on the subdomain, you would be prepending your custom autoloader instance so that it has higher priority than the generated instance to find classes (you need to do that very early, because it must be done before the first usage of a class you want to allow overriding, as PHP does not allow unloading classes).

**Reply by dadamssg** [reply]
Thanks @stof , something like this?
```php
// index.php

$root = dirname(dirname(__FILE__));

require_once $root . '/vendor/composer/ClassLoader';

$classLoader = new \Composer\Autoload\ClassLoader();
$subdomain = getSubdomainFromGlobals();
$classLoader->setPsr4('MyCompany\MyProject', $root . "/subdomain/" . $subdomain);
$classLoader->register();

require_once $root . '/vendor/autoload.php';

// ...

```
